fix: Исправлены ложные срабатывания диагностик MagicNumber и MagicDate при работе со структурами

johnnyshut opened the pull request:
## Описание
Исправлены ложные срабатывания диагностик MagicNumber и MagicDate при работе со структурами. Теперь диагностики не срабатывают при присваивании значений к свойствам структуры, так как имя свойства уже дает смысл значению.

## Связанные задачи
Closes #3496

## Чеклист

### Общие

- [x] Ветка PR обновлена из develop
- [x] Отладочные, закомментированные и прочие, не имеющие смысла участки кода удалены
- [x] Изменения покрыты тестами
- [x] Обязательные действия перед коммитом выполнены (запускал команду gradlew precommit)

### Для диагностик

- [x] Описание диагностики заполнено для обоих языков (присутствуют файлы для обоих языков, для русского заполнено все подробно, перевод на английский можно опустить)

## Дополнительно



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Configurable authorized dates and improved recognition of correspondence/map types; new shared detection base for magic values.

* **Bug Fixes**
  * Fewer false positives: diagnostics now skip default-value contexts, structure/correspondence usages, and certain index-access cases.

* **Refactor**
  * Consolidated magic-number/date detection into a common diagnostic base and streamlined eligibility checks.

* **Documentation**
  * Added "Exceptions" sections with examples for magic-number and magic-date rules.

* **Tests**
  * Updated tests, constructors, and expected diagnostic counts/ranges to align with refined detection.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nixel2007 commented on src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/MagicDateDiagnostic.java:
```diff
@@ -93,15 +105,23 @@ public ParseTree visitConstValue(BSLParser.ConstValueContext ctx) {
+      if (!insideSimpleDateAssignment(expressionContext) 
+          && !insideReturnSimpleDate(expressionContext)
+          && !insideAssignmentWithDateMethodForSimpleDate(expressionContext)
+          && !expressionContext.map(MagicDateDiagnostic::insideStructurePropertyAssignment).orElse(false)) {
```
Лучше соблюсти такой же стиль проверки условия, как и в трех проверках выше

johnnyshut commented on src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/AbstractVisitorDiagnostic.java:
```diff
@@ -47,4 +49,21 @@ public List<Diagnostic> getDiagnostics(DocumentContext documentContext) {
+   * @param ctx контекст узла для поиска
+   * @return контекст присваивания или null, если не найден
+   */
+  protected static BSLParser.AssignmentContext findAssignmentContext(ParserRuleContext ctx) {
```
`Trees.getRootParent(bslContext, Collections.singleton(BSLParser.RULE_assignment))`
Вот так сделал

johnnyshut commented on src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/MagicDateDiagnostic.java:
```diff
@@ -159,10 +320,63 @@ private static Optional<BSLParser.ExpressionContext> getExpression(Optional<BSLP
+        return assignmentText.contains("Дата(") && assignmentText.contains(")");
+      })
+      .orElse(false);
+  }
```
Кролик, проверь, исправил

nixel2007 commented:
@johnnyshut можешь над кавереджом в magic date поработать? Такое чувство, что visit String больше не вызывается, все обрабатывается на уровне visitConstValue с ранним возвратом.

johnnyshut commented:
> @johnnyshut можешь над кавереджом в magic date поработать? Такое чувство, что visit String больше не вызывается, все обрабатывается на уровне visitConstValue с ранним возвратом.

Да, Ты прав, хочу убрать лишнее и тесты плохо сделал

johnnyshut commented:
@nixel2007 готово, посмотри, пожалуйста

johnnyshut committed:
feat(diagnostics): Добавлен класс AbstractMagicValueDiagnostic и проведен рефакторинг существующих диагностик.

Введен AbstractMagicValueDiagnostic в качестве базового класса для обработки диагностик магических значений, инкапсулирующих общую логику проверки выражений внутри структур и соответствий. MagicDateDiagnostic и MagicNumberDiagnostic обновлены для расширения этого нового класса, упрощения их реализации и повышения возможности повторного использования кода.

Ключевые изменения включают:
- Создание AbstractMagicValueDiagnostic с методами для обработки контекста выражений.
- Рефакторинг MagicDateDiagnostic и MagicNumberDiagnostic для использования новых методов базового класса, что повышает удобство поддержки и ясность кода.
- Корректировки диагностических проверок для использования общей функциональности, предоставляемой AbstractMagicValueDiagnostic.

johnnyshut committed:
test: рефакторинг тестов диагностик и добавление тестов для структур

- Автоматическое определение класса диагностики через рефлексию в AbstractDiagnosticTest
- Удаление явных конструкторов из тестов MagicDateDiagnostic и MagicNumberDiagnostic
- Добавление тестовых случаев для проверки структур и соответствий
- Обновление ожидаемых количеств диагностик в тестах

johnnyshut committed:
doc: добавлены разделы об исключениях для магических дат и чисел

Обновлены файлы документации MagicDate.md и MagicNumber.md, добавлены разделы об исключениях, поясняющие, что магические даты и числа, используемые в структурах и соответствиях, не считаются ошибками. Приведены примеры использования в контексте структур данных.

johnnyshut committed:
refactor(diagnostics): замена магических чисел на константы

Вынесены магические числа (10, 50, 5) в именованные константы для улучшения читаемости кода и устранения замечаний линтера.

nixel2007 committed:
Merge branch 'develop' into HEAD
